File: Sort by Date/002 - June24 - Sept 5 2017 round mp/Deletion Distance.cs
using System;

class Solution
{
    public static int DeletionDistance(string str1, string str2, int index1, int index2, int[][] memo)
    {
        // your code goes here

        if ((str1 == null || str1.Length == 0) &&
           (str2 == null || str2.Length == 0))
        {
            return 0;
        }

        if (str1 == null || str1.Length == 0)
        {
            return str2.Length;
        }

        if (str2 == null || str2.Length == 0)
        {
            return str1.Length;
        }

        if (index1 == str1.Length)
        {
            return str2.Length - index2;
        }

        if (index2 == str2.Length)
        {
            return str1.Length - index1;
        }

        int length1 = str1.Length;
        int length2 = str2.Length;

        if (memo == null)  // memo "dog", "frog"
        {
            memo = new int[length1][];
            for (int i = 0; i < length1; i++)
            {
                memo[i] = new int[length2];
            }

            for (int row = 0; row < length1; row++)
            {
                for (int col = 0; col < length2; col++)
                {
                    memo[row][col] = -1;
                }
            }
        }

        bool noMemo = memo[index1][index2] == -1;

        if (!noMemo)
        {
            return memo[index1][index2];
        }

        // recursive call 
        var isEqual = str1[index1] == str2[index2];
        int distance = 0;
        if (isEqual)
        {
            distance = DeletionDistance(str1, str2, index1 + 1, index2 + 1, memo); // delete, 0 
        }
        else
        {
            var option1 = 1 + DeletionDistance(str1, str2, index1 + 1, index2, memo);  // delete one char
            var option2 = 1 + DeletionDistance(str1, str2, index1, index2 + 1, memo);  // delete one char
            distance = Math.Min(option1, option2);
        }

        memo[index1][index2] = distance;

        return distance;
    }

    static void Main(string[] args)
    {
        Console.WriteLine(DeletionDistance("dog", "frog", 0, 0, null));
        Console.WriteLine(DeletionDistance("some", "thing", 0, 0, null));

    }
}

//dog -> frog 
// heat -> hit , h -> eat vs it , e != i, 
// remove e, dist("at","it")
// remove i, dist("eat","t")
// Math.Min(dist("at","it"), dist("eat","t")), memoization key string "at it", "it at"
// left -> right, index , index -> unique -> int[] key int[2], compare function 
//memo[i][j] jagged array i, j already store
// base case: 
// "", "abc" 

File: 2018 02 jianminchen.2018apple January 24 - Feb 6 2018/Being an interviwer/DeletionDistance.java
import java.io.*;
import java.util.*;

class Solution {

  static int deletionDistance(String str1, String str2) {
    str1 = "$" + str1;
    str2 = "$" + str2;
    int m = str1.length();
    int n = str2.length();
    int[][] f = new int[m][n];
    for(int i = 0; i < m ; i++) {
      Arrays.fill(f[i], m + n + 1);
    }
    for(int i = 0; i < m; i++) {
      f[i][0] = i;
    }
    for(int i = 0; i < n; i++) {
      f[0][i] = i;
    }        
    
    for(int i = 1; i < m; i++) {
      for(int j = 1; j < n; j++) {
        if (str1.charAt(i) == str2.charAt(j)) {
          f[i][j] = f[i-1][j-1];
        } else {
          f[i][j] = Math.min(f[i-1][j],f[i][j-1]) + 1;
        }
      }
    }
    return f[m-1][n-1];
  }

  //$dog and $frop, 
  //$hit, i =   3 $heat, j = 4
  //f[$hit][$heat] = f[$hi][$hea]
  //(f[$hit][$hea] + 1)
  //(f[$hi][$heat] + 1)
  //(f[$hi][$hea]  <- minimum 
  //    min(f[$h][$hea], $f[$hi][$he]) + 1
  //)
  //f[3][4] = f[2][3]       
  
  //f[4][5]
  //f[0][0] = 0
  //f[i][j] 
    // i == j f[i][j] = f[i-1][j-1]
    // i !== f[i][j] = min(f[i-1][j],f[i][j-1]) + 1
  
  //f[0][0] = 0
  //f[0][1] = 1
  //f[0][2] = 2
  //...
  //f[1][1] = min(f[1][0], f[0][1]) + 1 = 2
  //...
  
  
  public static void main(String[] args) {
    System.out.println(deletionDistance("dog","frog"));
    System.out.println(deletionDistance("some","some"));
    System.out.println(deletionDistance("some","thing"));
  }
}